### DISCUSSION TITLE
Can i whitelist urls on https://dashboard.web3auth.io programatically?

### DISCUSSION THREAD

**Discussion by himanshuchawla009** [Web3Auth/Web3Auth#351] [Help]
- Is there any way to whitelist urls in developer dashboard from my backend?
- I am building a multi tenant application where i need to add a new sub domain for each my client whenever they signup in my app. Can i do it directly from my code?

**Comment by chaitanyapotti** [comment]
No need to whitelist on developer dashboard.
You can use manual origin whitelisting.

`OpenLoginAdapter` of web3auth accepts a parameter called `OriginData`.
`OriginData` is a key value pair
key: origin (e.g: https://a.example.com)
value: `signature`
signature is generated by using [`whitelistUrl`](https://github.com/torusresearch/OpenLoginSdk/blob/master/packages/openlogin/src/utils.ts#L26) function from the package [`@toruslabs/openlogin`](https://www.npmjs.com/package/@toruslabs/openlogin) package

The parameters for whitelistUrl function are (clientId, clientSecret, origin)

**Reply by himanshuchawla009** [reply]
can you please provide a code snippet for generating signatures in the format supported by web3auth.

**Reply by chaitanyapotti** [reply]
To use the same function in the backend and generate the signature,

```js
const base64url = require("base64url");
const keccak = require("keccak");
const { getPublic, sign } = require("@toruslabs/eccrypto");

const clientId = "BG0djgWQ8-sDz3OfdJmuQnO3XnsO8VOBhNGOSW8BgmgiNiit1374bM89r3RT0XP2n2BNkTS6A33UkQryCfkGbBs";
const clientSecret = "f785b7a8ef9e50bd878180c4b89072d08e8c2c59840bc01ef080f5b00783264e";
const origin = "https://www.example.com";
const appKeyBuf = Buffer.from(clientSecret.padStart(64, "0"), "hex");
if (base64url.encode(getPublic(appKeyBuf)) !== clientId) throw new Error("appKey mismatch");
const sig = await sign(appKeyBuf, Buffer.from(keccak("keccak256").update(origin).digest("hex"), "hex"));
const finalSig = base64url.encode(sig);
console.log("final sig", finalSig);
```

**Comment by billyjacoby** [comment]
The solution posted here is great if you aren't just running client side - but this doesn't allow me to use a wild card domain if i have a client side only application since I can't expose the secret there.

Is there still no way to add a wildcard domain to whitelist settings?

**Comment by marcus-sa** [comment]
Will wild card domains ever be supported..? This is ridiculous ...

**Comment by Unique-Divine** [comment]
Took some tinkering, but I got this to work in 2026 using the pattern originally described in [the above answer from @chaitanyapotti](https://github.com/orgs/Web3Auth/discussions/351#discussioncomment-3303780). 

# Signed `originData` for ephemeral Web3Auth preview domains

## Problem

Our application deploys a unique preview hostname for each pull request, for example:

```text
https://1051.gh-pr.sai.fun
```

The Web3Auth/MetaMask Embedded Wallets dashboard accepts an exact hostname allowlist entry, but its documented wildcard entry:

```text
https://*.gh-pr.sai.fun
```

was not honored for this project. Social login succeeded only after manually adding the exact numbered preview origin. Manually updating the dashboard for every pull request is not a practical or safe deployment workflow.

## Solution

Web3Auth v9 accepts a signed origin map through `AuthAdapter` settings:

```ts
new AuthAdapter({
  adapterSettings: {
    originData: {
      "https://1051.gh-pr.sai.fun": "<signature>",
    },
  },
})
```

The application does not need to mutate the dashboard allowlist. Instead, trusted build infrastructure generates a signature for the exact preview origin, exposes only that signature to the browser, and passes it to the SDK as `originData`.

The browser-visible `originData` value is scoped to one exact origin. It is not the Web3Auth client secret and should not authorize a different PR hostname or a wildcard domain.

## Trust boundary

The client secret remains in trusted CI only. In our case, GitHub Actions reads one organization secret named `WEB3AUTH_ORG_CLIENTS`:

```json
{
  "prodSai": {
    "clientId": "<mainnet Web3Auth client ID>",
    "clientSecret": "<mainnet Web3Auth client secret>"
  },
  "testSai": {
    "clientId": "<testnet Web3Auth client ID>",
    "clientSecret": "<testnet Web3Auth client secret>"
  }
}
```

The build output contains only this public shape:

```json
{
  "prod": {
    "domain": "https://1051.gh-pr.sai.fun",
    "web3AuthOriginDataSignature": "<DER ECDSA signature, base64url encoded>"
  },
  "test": {
    "domain": "https://1051.gh-pr.sai.fun",
    "web3AuthOriginDataSignature": "<DER ECDSA signature, base64url encoded>"
  }
}
```

That JSON is assigned to `NEXT_PUBLIC_WEB3AUTH_ORIGIN_DATA` at build time. No `NEXT_PUBLIC_*` variable contains a client secret.

For safety, the secret-bearing workflow step runs only for same-repository pull requests. Forked PRs do not receive the organization secret.

## Signature format

The implementation reproduces the OpenLogin/Web3Auth `whitelistUrl` algorithm:

1. Interpret `clientSecret` as a secp256k1 private key, padded to 32 bytes.
2. Derive the uncompressed public key.
3. Base64url-encode that public key and require it to equal `clientId`. This prevents signing with the wrong project secret.
4. Compute `keccak256(utf8(origin))` for the exact origin string.
5. Create a secp256k1 ECDSA signature of that hash.
6. Convert the signature to DER encoding.
7. Base64url-encode the DER bytes for the `originData` map value.

The DER step matters. A compact 64-byte `r || s` signature is not equivalent to the OpenLogin helper output. We added a conformance test against OpenLogin's own `whitelistUrl` function and used it to catch this distinction.

## Server-side signing script

This is the sanitized TypeScript implementation. It uses placeholders only; it must run in CI or another trusted server environment, never in browser code.

```ts
import { keccak256 } from "ethereum-cryptography/keccak.js"
import {
  ecdsaSign,
  publicKeyCreate,
  signatureExport,
} from "ethereum-cryptography/secp256k1-compat.js"
import { hexToBytes, utf8ToBytes } from "ethereum-cryptography/utils.js"

type Web3AuthClient = {
  clientId: string
  clientSecret: string
}

type Web3AuthClients = {
  prodSai: Web3AuthClient
  testSai: Web3AuthClient
}

function exactHttpsOrigin(value: string): string {
  const parsed = new URL(value)
  if (parsed.protocol !== "https:" || parsed.origin !== value) {
    throw new Error("origin must be an exact HTTPS origin without a path")
  }
  return value
}

function privateKey(clientSecret: string): Uint8Array {
  const hex = clientSecret.replace(/^0x/, "")
  if (!/^[0-9a-f]{1,64}$/i.test(hex)) {
    throw new Error("clientSecret must be a hexadecimal secp256k1 private key")
  }
  return hexToBytes(hex.padStart(64, "0"))
}

function signOrigin(
  client: Web3AuthClient,
  origin: string,
): string {
  const key = privateKey(client.clientSecret)

  // `false` requests the uncompressed 65-byte key expected by OpenLogin.
  const publicKey = publicKeyCreate(key, false)
  if (Buffer.from(publicKey).toString("base64url") !== client.clientId) {
    throw new Error("clientId does not match clientSecret")
  }

  const hash = keccak256(utf8ToBytes(origin))
  const compactSignature = ecdsaSign(hash, key).signature

  // OpenLogin `whitelistUrl` encodes a DER signature, not compact r || s bytes.
  return Buffer.from(signatureExport(compactSignature)).toString("base64url")
}

export function generateOriginData(
  clients: Web3AuthClients,
  previewOrigin: string,
) {
  const domain = exactHttpsOrigin(previewOrigin)

  return {
    prod: {
      domain,
      web3AuthOriginDataSignature: signOrigin(clients.prodSai, domain),
    },
    test: {
      domain,
      web3AuthOriginDataSignature: signOrigin(clients.testSai, domain),
    },
  }
}
```

## GitHub Actions integration

The workflow already knows the final preview URL from the pull request number. A representative CI step is:

```yaml
- name: Generate signed Web3Auth preview origin data
  if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
  env:
    WEB3AUTH_ORG_CLIENTS: ${{ secrets.WEB3AUTH_ORG_CLIENTS }}
  run: |
    bun scripts/generate-web3auth-origin-data.ts \
      "https://${{ github.event.pull_request.number }}.gh-pr.sai.fun" \
      >> "$GITHUB_ENV"
```

The script writes one `GITHUB_ENV` assignment:

```text
NEXT_PUBLIC_WEB3AUTH_ORIGIN_DATA={...public signatures only...}
```

The subsequent Next.js build receives that public variable and embeds it in the preview bundle.

## Browser-side adapter wiring

The webapp parses the public JSON, selects `prod` for mainnet and `test` otherwise, and converts the selected entry into the SDK's required map:

```ts
type PublicOriginData = {
  prod?: {
    domain: string
    web3AuthOriginDataSignature: string
  }
  test?: {
    domain: string
    web3AuthOriginDataSignature: string
  }
}

function adapterOriginData(
  value: string | undefined,
  isMainnet: boolean,
): Record<string, string> | undefined {
  if (!value) return undefined

  try {
    const parsed = JSON.parse(value) as PublicOriginData
    const entry = isMainnet ? parsed.prod : parsed.test
    if (!entry?.domain || !entry.web3AuthOriginDataSignature) return undefined

    const origin = new URL(entry.domain)
    if (origin.protocol !== "https:" || origin.origin !== entry.domain) {
      return undefined
    }

    return { [entry.domain]: entry.web3AuthOriginDataSignature }
  } catch {
    return undefined
  }
}

const originData = adapterOriginData(
  process.env.NEXT_PUBLIC_WEB3AUTH_ORIGIN_DATA,
  isMainnet,
)

const authAdapter = originData
  ? new AuthAdapter({ adapterSettings: { originData } })
  : new AuthAdapter()
```

The fallback is intentional: local development and non-preview deployments keep their existing dashboard-driven behavior when the public payload is absent or malformed.

## Versions used

The application uses the following Web3Auth v9 packages, resolved at `9.7.0`:

- `@web3auth/auth-adapter@^9.7.0`
- `@web3auth/base@^9.7.0`
- `@web3auth/base-evm-adapter@^9.7.0`
- `@web3auth/ethereum-provider@^9.7.0`
- `@web3auth/modal@^9.7.0`
- `@web3auth/no-modal-react-hooks@^9.7.0`

The signer uses `ethereum-cryptography@3.2.0`. The test-only reference implementation is `@toruslabs/openlogin@8.2.1`, whose `whitelistUrl(clientId, clientSecret, origin)` output is compared against two non-sensitive fixtures.

## Validation strategy

Automated tests should cover:

- exact-origin validation: reject paths, query strings, fragments, non-HTTPS URLs, and normalized-but-different origins;
- missing or malformed CI secret JSON;
- a client ID that does not match its client secret;
- `prod` and `test` output generation;
- no client secret in the generated public JSON;
- conformance with OpenLogin's `whitelistUrl` output, including DER encoding;
- browser payload parsing and the no-payload fallback.

The final integration test is manual because social OAuth requires an interactive browser session:

1. Deploy a same-repository pull request preview.
2. Remove the exact numbered preview hostname from the relevant Web3Auth dashboard project.
3. Open the preview in a private browser session.
4. Complete a social login.
5. Verify the resulting embedded wallet can sign.

A successful login with no exact dashboard entry demonstrates that the signed `originData` path, rather than the dashboard wildcard, authorized the preview origin.

## Security notes

- Do not log, commit, paste, or expose `clientSecret` values.
- Do not put client secrets in a `NEXT_PUBLIC_*` variable.
- Sign the exact `window.location.origin` value: no trailing slash, path, query, fragment, wildcard, or case variation.
- A signature should be treated as public but origin-scoped authorization data.
- Restrict secret-bearing CI steps to trusted pull requests.
- Keep the dashboard wildcard issue tracked separately; this signed-origin design is a practical workaround, not evidence that wildcard allowlisting works.
